### PR TITLE
Add 2fa authentication

### PR DESCRIPTION
**Pull request by OscBacon**

### CODE REVIEW

File: src/core/Directus/Authentication/Exception/TFAEnforcedException.php [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+use Directus\Exception\UnauthorizedException;
+
+class TFAEnforcedException extends UnauthorizedException
+{
+    const ERROR_CODE = 113;
+
+    public function __construct()
+    {
+        parent::__construct('2FA enforced but not activated for user');
```
Comment (rijkvanzanten) [inline src/core/Directus/Authentication/Exception/TFAEnforcedException.php]: Does this need a second parameter? The error above looks like

`parent::__construct('User missing 2FA OTP', static::ERROR_CODE);`

but this one is missing `static::ERROR_CODE`

**Comment by rijkvanzanten** [comment]
Code LGTM.

File: src/core/Directus/Authentication/Exception/TFAEnforcedException.php (hunk 1 above)
Comment (OscBacon) [reply src/core/Directus/Authentication/Exception/TFAEnforcedException.php]: That parameter isn't needed to extend an UnauthorizedException

File: composer.json [hunk 2]
```diff
@@ -39,6 +39,7 @@
     "ext-mbstring": "*",
     "ext-openssl": "*",
     "ext-gd": "*",
+    "phpgangsta/googleauthenticator": " dev-master",
```
[the remark this answers is not in the document]
Comment (OscBacon) [reply composer.json]: Different library was used

File: src/core/Directus/Authentication/Exception/Missing2FAPasswordException.php [hunk 3]
```diff
@@ -0,0 +1,15 @@
+<?php
+
+namespace Directus\Authentication\Exception;
+
+use Directus\Exception\NotFoundException;
+
+class Missing2FAPasswordException extends NotFoundException
+{
+    const ERROR_CODE = 100;
```
[the remark this answers is not in the document]
Comment (OscBacon) [reply src/core/Directus/Authentication/Exception/Missing2FAPasswordException.php]: Changed error code to 111

**Comment by rijkvanzanten** [comment]
@OscBacon Do you mind also writing some docs on how the feature works? 🙂

**Comment by rijkvanzanten** [comment]
Ohh shit, there's 1 huge problem with this PR: you can't login anymore without running the DB upgrade 😨

The API tries to read the 2fa columns, but they don't exist yet. To make them show up, you have to run the migrations, but you can't make it to settings, because the app can't load:

image

This means that people will be locked out...... can we think of a way to make this somewhat backwards compatible?

@bjgajjar @benhaynes  Can we / should we come up with a way for the API to automatically run the database migrations on first load after a pull? If that's too risky, can we come up with a way for the API to let the application know that the database needs to be upgraded? So far, the DB upgrades have only changed non-critical things like default interfaces or field widths, but this one in particular kinda is a breaking change

**Comment by OscBacon** [comment]
> Ohh shit, there's 1 huge problem with this PR: you can't login anymore without running the DB upgrade 
> 
> The API tries to read the 2fa columns, but they don't exist yet. To make them show up, you have to run the migrations, but you can't make it to settings, because the app can't load:
> 
> image
> 
> This means that people will be locked out...... can we think of a way to make this somewhat backwards compatible?
> 
> @bjgajjar @benhaynes Can we / should we come up with a way for the API to automatically run the database migrations on first load after a pull? If that's too risky, can we come up with a way for the API to let the application know that the database needs to be upgraded? So far, the DB upgrades have only changed non-critical things like default interfaces or field widths, but this one in particular kinda is a breaking change

What about checking if the 2fa columns exist in functions that use 2fa? 
That way we can just ignore 2fa capabilities if they aren't present in the database?

Also, I'll work on the docs tomorrow!

**Comment by bjgajjar** [comment]
Hey @OscBacon 

The code looks okay to me too. But I think instead of adding `enforce_2fa` in `directus_user_roles` we need to add it in `directus_settings` table. 

We just need to check that whether the 2FA is enabled at the site level or not. If it enables at site level then the user will get the option to enable it for them self too. 

I am not clear with the use case to enable it as a role specific. Can you please highlight it?

Hey @rijkvanzanten 

We can resolve this issue in 2 ways.

1. We can update the DB using migration command.
2. We have one option `Version and Updates` in the `Admin Setting`. From here we can update the build to the latest one as well as migrate the DB too which we need to develop.

**Comment by rijkvanzanten** [comment]
>  But I think instead of adding enforce_2fa in directus_user_roles we need to add it in directus_settings table

I don't agree. As admin, you might not need / want to enforce 2fa for literally every user. For example, I would enforce 2fa for all `admin` users, but leave 'regular' content moderator roles as is, so less tech savvy users don't have to use 2FA. This is less of a security issue, as their permissions are severely restricted anyway.

> We can update the DB using migration command.
> We have one option Version and Updates in the Admin Setting. From here we can update the build to the latest one as well as migrate the DB too which we need to dev

The second one sounds like the best way forward. Have a button in the application (+ endpoint in the API) that will update the API and run the migrations. The only problem we have then is how to update the application. If the application is on the same server as the API, the API can update the app. If it's not, we have a problem

**Comment by benhaynes** [comment]
Hey @bjgajjar  — @OscBacon had originally set this to Global (`directus_settings`) but since our auth tokens are all based on users, you might have some roles that are used to connect to apps... and can't have 2fa. So we decided that attaching this to `directus_roles` would be more granular.

Does that make sense?

**Comment by bjgajjar** [comment]
@rijkvanzanten @benhaynes 

Okay, Let me clear both the scenarios here.

(1) Add role specific 2FA (`directus_user_role`)

2FA will be enabled based on **the role**. If admin enables it only for `admin`, then only `admin` will get the option of 2FA in their profile. They can enable/disable from there. Except `admin` none of any users will get the option for this.

(2) Add 2FA as a site level (`directus_settings`)

From the global setting; `admin` will set that the 2FA is enabled/disabled for the **the whole site**. If it is enabled then all the users(contains all roles) will get the option to enable/disable the 2FA from their profile; otherwise, none of them will able to do so.

> I would enforce 2fa for all admin users, but leave 'regular' content moderator roles as is, so less tech-savvy users don't have to use 2FA. 

**P.S.** When admin enables it for a **site level** it doesn't mean that it will enable for all the system users automatically. User needs to enable it for themselves manually from their profiles. The same flow will go with the role-specific condition. When 2FA enable for any role, then it will not enables 2FA for the users of that role automatically. They need to enable it from their profile only as all the users have their own 2FA secret.

Let me know your thoughts.

**Comment by rijkvanzanten** [comment]
> 2FA will be enabled based on the role. If admin enables it only for admin, then only admin will get the option of 2FA in their profile. They can enable/disable from there. Except admin none of any users will get the option for this.

No, everyone has the ability to turn on 2FA. The toggle in the role settings is _enforcing_ it. EG the user _has to have_ 2fa enabled.

I think you misunderstood the role / site level 2FA _enforcement_

**Comment by OscBacon** [comment]
Hey @bjgajjar @benhaynes @rijkvanzanten 
We still have the incompatibility issue where if someone doesn't migrate the db, they will run into issues. Do you have a final thought on how to fix this?

**Comment by rijkvanzanten** [comment]
I think we should add a check that either checks availability of this column based on the latest run migration, or based on the existence of the column itself, instead of trying to read it and fail.

Ideally, we can have a way to 'force' the user into a database upgrade on login, but seeing that that's a big new feature, it'll probably take too long to implement to have this wait on that.

**Comment by bjgajjar** [comment]
Hey @rijkvanzanten,

I guess now its a time to develop this feature where the user will not able to use neither `API` nor `APP` if they have any migration which has been not executed yet. Maybe it will take some time but it will be worth it. As without this feature, we are not able to merge our major functionality which has been built; `2FA`.

- From the `API`, one particular status code will be sent along with an error message if there is any migration which is not executed yet.

- In the `APP`, if that status code had returned from `API` then the user will be redirected to that page from where they can migrate DB. After the migration process has been completed, the user will able to use the `APP`. 

Let me know your thoughts here. So we can start the work :)

**Comment by OscBacon** [comment]
What we could do is that in the Phinx/Migration/Manager.php, there is a way to compare what has already been migrated (versions) to the migrations available (migrations)

Why not use that and make it so that if versions != migrations, we send an error ?
image

**Comment by rijkvanzanten** [comment]
@bjgajjar You are right. I just didn't know how long it would take!

Lets discuss this further in https://github.com/directus/api/issues/1107

**Comment by benhaynes** [comment]
If we're redirecting, then it might be good to have this App page completed:

https://github.com/directus/app/issues/1370

That way they can get more information (beyond just clicking a button). Also, I worry about small migrations causing API downtime — especially for small migrations that wouldn't normally affect the API responses (like removing a field from users). Could we still return API responses like normal, but still have the error message let the user/app know what needs to happen?

Or maybe I'm overthinking it and we should just force them. 🤔

**Comment by rijkvanzanten** [comment]
@benhaynes The API will continue to work as normal*

* Unless the migrations are a breaking change, as is the case with 2FA

**Comment by OscBacon** [comment]
@bjgajjar @benhaynes  :)

**Comment by benhaynes** [comment]
We haven't forgotten about this one, @OscBacon — just quite busy lately and haven't had time to properly review the front-end. :)

**Comment by bjgajjar** [comment]
@benhaynes, 

I am checking both PR of 2FA in `APP` as well as `API`. There were some issues and I recommended changes to @OscBacon  which he pushed in the evening. 

I'll review it ASAP and send him feedback.

**Comment by OscBacon** [comment]
Thanks @benhaynes and @bjgajjar :)

**Comment by bjgajjar** [comment]
@OscBacon 

Still, we have one issue here. 

When disable the enforce 2FA from the particular role; the `2fa_secret` field of `directus_users` remains the same(Not Null). That's why getting the screen of a passcode after login although admin disables the enforcing for that role.

Kindly set `2fa_secret` field of `directus_users` as `null` when disable the enforce 2FA from the particular role.

As well as the changes of  `src/core/Directus/Services/ItemsService.php` file are unwanted. You must remove that code as I replaced it already in older PR.  Kindly take care when you are taking the pull and resolve any conflicts.

**Comment by OscBacon** [comment]
Hi @bjgajjar,
"enforce" means the user should be forced to use 2fa. The idea is that if 2fa is not enforced, then 2fa is optional for that user. If a user has activated 2fa, shouldn't they keep their 2fa?

@kaushal-im any thoughts on this too?

**Comment by bjgajjar** [comment]
> If a user has activated 2fa, shouldn't they keep their 2fa?

Agree with you here :) Yeah that's not an issue! Just remove the code. Will merge the PR :)

**Comment by kaushal-im** [comment]
@OscBacon Seems perfect. Thanks for your contribution ❤

**Comment by OscBacon** [comment]
Thanks @kaushal-im, glad it all came together! 😊

**Pull request merged**